### ISSUE TITLE
change boundQueue process from *span to []*span to less channel concurrency.

### ISSUE THREAD

**Issue by zdyj3170101136** [jaegertracing/jaeger#3665]
## Requirement - what kind of business use case are you trying to solve?
I am using collector to receive data through grpc from agent and use json.Marshal send it to kafka.
The problem is the collector server with 10% cpu usage, but high span drop rate.
Turn up bound queue buffer size seems helpless.

## Problem - what in Jaeger blocks you from solving the requirementation
Jaeger collector boundQueue do not support batch process []*span.

## Proposal - what do you suggest to solve the problem or improve the existing situation?
Change the boundQueue process from *span to []*span.
By default, the boudQueue send *span to chan, and the worker read *span from chan and process the *span.
```
queue := make(chan interface{}, capacity)
	return &BoundedQueue{
		onDroppedItem: onDroppedItem,
		items:         &queue,
		stopCh:        make(chan struct{}),
		capacity:      uatomic.NewUint32(uint32(capacity)),
		stopped:       uatomic.NewUint32(0),
		size:          uatomic.NewUint32(0),
	}
``` 
I think the chan is the bottelneck.
It should change the queue.Produce func from receive *span to  []*span to less concurrent write and read frequency.
```
queue := make(chan []interface{}, capacity)
	return &BoundedQueue{
		onDroppedItem: onDroppedItem,
		items:         &queue,
		stopCh:        make(chan struct{}),
		capacity:      uatomic.NewUint32(uint32(capacity)),
		stopped:       uatomic.NewUint32(0),
		size:          uatomic.NewUint32(0),
	}
``` 
And we split []*span to [][maxBufferSize]*span in spanProcessor ProcessSpans.
```
const (
   maxBufferSize = 50
)

func splitSpans(spans []*model.Span) [][]*model.Span {
   var res [][]*model.Span
   tmp := make([]*model.Span, 0, maxBufferSize)
   for _, span := range spans {
      tmp = append(tmp, span)
      if len(tmp) == maxBufferSize {
         res = append(res, tmp)
         tmp = make([]*model.Span, 0, maxBufferSize)
      }
   }
   return res
}
```
This will refactor many code to support batch process *span.
But after using it in prod, the span drop rate is lower, almost zero.
## Any open questions to address
My question is how should i choose the right number of maxBufferSize.
With lower number, the chan concurrency is higher.
With higher number, some agent with many trace data may be block other service's trace data, and thus let collector drop other service's trace data.

**Comment by yurishkuro** [comment]
What is your evidence that this aspect of the queue has anything to do with CPU usage or dropped spans?

**Comment by zdyj3170101136** [comment]
After check the grafana, i find that the queue length is even higher after i change it to batch queue.
So maybe i was wrong.
I will try to republish the code with a no batch queue version, and compare it.

**Comment by zdyj3170101136** [comment]
I made a test use prod environment.
My machine is 12 core, 24 gb memory, amd64.
I run collector with param  --collector.num-workers=12,--collector.queue-size=1000000 and env GOGC=1000.

Here is the grafana metric of collector which run without batch queue.
It receive about 5w span/s,  it's span drop rate about 5% temporarily.
And the cpu usage is only 1.7 core, which is about 15%  of machine.
With low cpu usage, i think is not a great idea to upgrade machine to get lower span drop rate.
<img width="807" alt="截屏2022-05-10 下午3 33 53" src="https://user-images.githubusercontent.com/52686065/167574128-7011e863-a60e-46ee-9b27-af2826a77ac4.png">

<img width="1627" alt="截屏2022-05-10 下午3 27 29" src="https://user-images.githubusercontent.com/52686065/167572969-209bfa66-7f72-4966-bd69-ca9fcebf9207.png">

<img width="1623" alt="截屏2022-05-10 下午3 30 32" src="https://user-images.githubusercontent.com/52686065/167573521-6989dee7-4f39-4524-a1e1-40d65946b99c.png">

Here is the grafana metric of collector which run with 50 size of  batch queue.
It receive about 5w span/s,  it's span drop rate about 0%.
And the cpu usage is only 1.8 core, which is about 18%  of machine.

Compare with the first situation, the span queue length is descend from 96K to only 50.
And the span queue time is descend from 3 to 0.075.
So Obviously batch queue is highly improve the performance.

<img width="1640" alt="截屏2022-05-10 下午3 37 49" src="https://user-images.githubusercontent.com/52686065/167574808-b88921fd-1e92-4253-89a7-d59afde46d64.png">
<img width="1638" alt="截屏2022-05-10 下午3 38 03" src="https://user-images.githubusercontent.com/52686065/167574850-cda1253c-6f04-473e-87b2-0fd25a2b1bda.png">

<img width="812" alt="截屏2022-05-10 下午3 38 59" src="https://user-images.githubusercontent.com/52686065/167575028-d812d330-2b0a-42ed-91c1-0c93f3874b47.png">

**Comment by yurishkuro** [comment]
Sorry, I don't really follow your experiment. Are you saying that you keep the queue capacity fixed (at 1M items), but vary the batch size of the items in the queue? The spikes in queue length are still at about 10% of the queue capacity, so that should not be causing span drops (unless your charts are not precise enough to capture max queue length).

If you're keeping the same queue capacity but allow batches in each queued item (say 50), then you're affectively configuring the collector with 50M queue capacity in terms of memory usage. So I think a comparable experiment would be:
  1. queue capacity=50M, batch size=1
  2. queue capacity=1M, batch size=50

WIth these settings I would expect your span drop rates to be similar between the runs.

Final point - the queue is not a bottleneck in the processing. Queue exists only to amortize certain short-lived spikes in traffic, by keeping spans in memory until the storage can write them. The actual bottleneck is the storage.

**Comment by zdyj3170101136** [comment]
Yes, i keeping the same queue capacity both in batch queue and not batch queue.

But i modify the behavior of Produce to let the queue.Size() means the real number span in the queue.
And the queue.Capiticy() is both 100 M.
In func queue.Produce, each items([]interface) from queue(chan []interface) will let the queue.Size add len(items).

I accept that 'the queue exists only to amortize certain short-lived spikes in traffic'.
But with batch queue, the consumer speed is faster, so i do not have to upgrade my machine to  get a zero span drop rate.

**Comment by yurishkuro** [comment]
I am not sure I follow your explanation. If you use the same _total_ queue capacity (i.e in case of batched queue you flatten the batch for the purpose of capacity calculation), then how do you explain that unbatched queue drops the spans? As I said, the only time it drops spans is when there's no more space in the queue, but if you claim that _total_ capacity is identical, then the situation can only arise simultaneously in both versions.
